Mathlib/Analysis/Calculus/Gradient.lean

```diff
@@ -0,0 +1,472 @@
+  rw [HasFDerivAt, HasFDerivAtFilter, Asymptotics.isLittleO_iff] at h
+  intro ε epos
+  specialize h epos
+  rw [Filter.Eventually] at h
+  let T := {u | ‖f u - f x - (f' x) (u - x)‖ ≤ ε * ‖u - x‖}
+  rcases (Iff.mp Metric.mem_nhds_iff h) with ⟨ε', ε'pos, h⟩
+  use (ε' / 2); constructor
+  · exact (half_pos ε'pos)
+  · intro x' xnhds
+    have : x' ∈ Metric.ball x ε':= by
+      rw [Metric.mem_ball, dist_comm]
+      rw [← dist_eq_norm] at xnhds
+      apply lt_of_le_of_lt xnhds (half_lt_self ε'pos)
+    have : x' ∈ T := h this
+    rw [Set.mem_setOf] at this
+    rw [norm_sub_rev x]
+    exact this
+
+theorem Convergence_HasFDeriv (h : ∀ ε > (0 : ℝ), ∃ δ > (0 : ℝ), ∀ (x' : E),
+    ‖x - x'‖ ≤ δ → ‖f x' - f x - (f' x) (x' - x)‖ ≤ ε * ‖x - x'‖) :
+      HasFDerivAt f (f' x) x := by
+  rw [HasFDerivAt, HasFDerivAtFilter, Asymptotics.isLittleO_iff]
+  intro ε epos
+  rw [Filter.Eventually]
+  specialize h ε epos
+  rcases h with ⟨δ, dpos, h⟩
+  rw [Metric.mem_nhds_iff]
+  use δ; constructor
+  apply dpos
+  intro x' x1mem
+  have : ‖x - x'‖ ≤ δ:= by
+    rw [Metric.ball, Set.mem_setOf, dist_comm, dist_eq_norm] at x1mem
+    exact LT.lt.le x1mem
+  rw [Set.mem_setOf, norm_sub_rev x']
+  apply h x' this
+
+/- The equivalent statement of `HasFDerivAt` in the `ε - δ` language-/
+theorem HasFDeriv_iff_Convergence_Point {gradient' : (E →L[ℝ] ℝ)}:
+  HasFDerivAt f (gradient') x ↔ ∀ ε > (0 : ℝ), ∃ δ > (0 : ℝ), ∀ (x' : E),
+     ‖x - x'‖ ≤ δ → ‖f x' - f x - (gradient') (x' - x)‖ ≤ ε * ‖x - x'‖ := by
+  constructor
+  apply HasFDeriv_Convergence
+  apply Convergence_HasFDeriv
+
+theorem HasFDeriv_iff_Convergence :
+  HasFDerivAt f (f' x) x ↔ ∀ ε > (0 : ℝ), ∃ δ > (0 : ℝ), ∀ (x' : E),
+     ‖x - x'‖ ≤ δ → ‖f x' - f x - (f' x) (x' - x)‖ ≤ ε * ‖x - x'‖ := by
+  constructor
+  apply HasFDeriv_Convergence
+  apply Convergence_HasFDeriv
+
+/- The equivalent statement of `HasGradnAt` in the `ε - δ` language-/
+theorem HasGradnAt_iff_Convergence: HasGradnAt f gradf x ↔
+  ∀ ε > (0 : ℝ), ∃ δ > (0 : ℝ), ∀ (x' : E),
+  ‖x - x'‖ ≤ δ → ‖f x' - f x - ⟪gradf, (x' - x)⟫‖ ≤ ε * ‖x - x'‖ := by
+  rw [HasGradnAt, HasFDeriv_iff_Convergence_Point]
+  rfl
+
+/- The following theorems are about the translation of `HasFDerivAt` and `HasGradnAt`-/
+theorem HasGradn_HasFDeriv  (h: HasGradnAt f gradient x):
+  HasFDerivAt f (∇* gradient) x := h
+
+theorem HasFDeriv_HasGradn (h: HasFDerivAt f gradient' x): HasGradnAt f (grad gradient') x := by
+  rw [HasGradnAt, HasFDeriv_iff_Convergence_Point]
+  rw [HasFDeriv_iff_Convergence_Point] at h
+  intro e epos
+  rcases (h e epos) with ⟨d, dpos, h⟩
+  use d; constructor; exact dpos
+  intro x' xle; specialize h x' xle
+  rw [← toDual_of_toDual_eq_self]
+  exact h
+
+theorem HasFDeriv_HasGradn' (h: HasFDerivAt f (∇* gradient) x):
+   HasGradnAt f gradient x := by rw [HasGradnAt]; exact h
+
+theorem HasGradn_HasFDeriv' (h: HasGradnAt f (grad gradient') x): HasFDerivAt f gradient' x := by
```

Comment:
```suggestion
theorem HasGradn_HasFDeriv' (h : HasGradnAt f (grad gradient') x) : HasFDerivAt f gradient' x := by
```

Reply:
Rename to use projection notation: `HasGradientAt.hasFDerivAt`.  Similar throughout.  (If you are new to Lean, you may need to study how projection notation works.)